synapse/rest/client/v1/room.py

```diff
@@ -267,6 +268,209 @@ def on_PUT(self, request, room_id, event_type, txn_id):
+        next_chunk_id = random_string(64)
+        insertion_event = {
+            "type": EventTypes.MSC2716_INSERTION,
+            "sender": requester.user.to_string(),
+            "content": {
+                EventContentFields.MSC2716_NEXT_CHUNK_ID: next_chunk_id,
+                EventContentFields.MSC2716_HISTORICAL: True,
+                # TODO: Why is `body` necessary for this to show up in /messages
+                "body": "TODO_REMOVE - INSERTION",
+            },
+            # Since the insertion event is put at the end of the chunk,
+            # copy the origin_server_ts from the last event we're inserting
+            "origin_server_ts": events_to_create[0]["origin_server_ts"],
+        }
+        # Prepend the insertion event to the start of the chunk
+        events_to_create = [insertion_event] + events_to_create
+
+        event_ids = []
+        prev_event_ids = prev_events_from_query
+        events_to_persist = []
+        for ev in events_to_create:
+            assert_params_in_dict(ev, ["type", "origin_server_ts", "content", "sender"])
+
+            # Mark all events as historical
+            # This has important semantics within the Synapse internals to backfill properly
+            ev["content"][EventContentFields.MSC2716_HISTORICAL] = True
+
+            event_dict = {
+                "type": ev["type"],
+                "origin_server_ts": ev["origin_server_ts"],
+                "content": ev["content"],
+                "room_id": room_id,
+                "sender": ev["sender"],  # requester.user.to_string(),
+                "prev_events": prev_event_ids.copy(),
+            }
+
+            event, context = await self.event_creation_handler.create_event(
+                requester,
+                event_dict,
+                prev_event_ids=event_dict.get("prev_events"),
+                # TODO: Do we need to use `self.auth.compute_auth_events(...)` to filter the `auth_event_ids`?
+                auth_event_ids=auth_event_ids,
+                inherit_depth=True,
+            )
+            current_state_ids = await context.get_current_state_ids()
+            logger.info(
+                "bulksend event=%s current_state_ids=%s",
+                event,
+                current_state_ids,
+            )
+
+            # TODO: Should we add the same `hs.is_mine_id(event.sender)` assert check that `create_and_send_nonmember_event` has?
+
+            events_to_persist.append((event, context))
+            event_id = event.event_id
+
+            event_ids.append(event_id)
+            # We add `event_id` so it references the last message.
+            # We add `prev_events_from_query` so it can find the proper depth
+            # while persisting. I wish we could rely on just `event_id` but
+            # since we are persisting in reverse-chronolical order below,
+            # that event isn't persisted yet.
+            prev_event_ids = [event_id] + prev_events_from_query
```

Comment:
Interesting problem where because we have to create events in chronological and persist in reverse-chronological, the `prev_events` don't exist yet so it has trouble finding successor depth. I added `prev_events_from_query` to the `prev_events` so it always knows where to go.

Any weird side-effects from having multiple `prev_events`? I think it's fine 🤔

Reply:
I wonder whether instead of `inherit_depth` we should pass through a `depth`? That way we can figure out the depth up here and just pass it through when creating each event?

I don't think having multiple prev events is particularly harmful in this case, just a bit confusing.

Reply:
> I wonder whether instead of `inherit_depth` we should pass through a `depth`

I think that makes sense :+1: Will update